iceoryx_dust/vocabulary/include/iox/detail/span_iterator.hpp

```diff
@@ -0,0 +1,223 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+#ifndef IOX_DUST_VOCABULARY_SPAN_ITERATOR_HPP
+#define IOX_DUST_VOCABULARY_SPAN_ITERATOR_HPP
+
+// Use 'assert's as 'iox::Expects' is not useable inside 'constexpr' functions
+#include <cassert>
+#include <iterator> // for reverse_iterator, distance, random_access_...
+
+namespace iox
+{
+
+template <typename T>
+class span_iterator
+{
+  public:
+    using iterator_category = std::random_access_iterator_tag;
+    using value_type = std::remove_cv_t<T>;
+    using difference_type = std::ptrdiff_t;
+    using pointer = T*;
+    using reference = T&;
+
+    constexpr span_iterator() = default;
+    constexpr span_iterator(const span_iterator& other) = default;
+    constexpr span_iterator(span_iterator&& other) noexcept = default;
+    constexpr span_iterator& operator=(const span_iterator& other) = default;
+    constexpr span_iterator& operator=(span_iterator&& other) noexcept = default;
+    ~span_iterator() = default;
+
+    constexpr span_iterator(pointer start, const pointer end)
+        : span_iterator(start, end, start)
+    {
+    }
+
+    constexpr span_iterator(const pointer begin, const pointer end, pointer current)
+        : m_begin(begin)
+        , m_end(end)
+        , m_current(current)
+    {
+    }
+
+    explicit constexpr operator span_iterator<const T>() const noexcept
+    {
+        return {m_begin, m_end, m_current};
+    }
+
+    constexpr reference operator*() const noexcept
+    {
+        assert(m_begin && m_end);
+        assert(m_begin <= m_current && m_current < m_end);
+        return *m_current;
+    }
+
+    constexpr pointer operator->() const noexcept
+    {
+        assert(m_begin && m_end);
+        assert(m_begin <= m_current && m_current < m_end);
+        return m_current;
+    }
+    constexpr span_iterator& operator++() noexcept
+    {
+        assert(m_begin && m_current && m_end);
+        assert(m_current < m_end);
+
+        ++m_current;
+        return *this;
+    }
+
+    // Rule DCL21-CPP is deprecated
+    // NOLINTNEXTLINE(cert-dcl21-cpp)
+    constexpr span_iterator operator++(int) noexcept
+    {
+        span_iterator ret = *this;
+        ++*this;
+        return ret;
+    }
+
+    constexpr span_iterator& operator--() noexcept
+    {
+        assert(m_begin && m_end);
+        assert(m_begin < m_current);
+        --m_current;
+        return *this;
+    }
+
+    // Rule DCL21-CPP is deprecated
+    // NOLINTNEXTLINE(cert-dcl21-cpp)
+    constexpr span_iterator operator--(int) noexcept
+    {
+        span_iterator ret = *this;
+        --*this;
+        return ret;
+    }
+
+    constexpr span_iterator& operator+=(const difference_type n) noexcept
+    {
+        if (n != 0)
+        {
+            assert(m_begin && m_current && m_end);
```

Comment:
Why doing this not unconditional? It feels weird to add and assign 0 to a `nullptr`. Alternatively an early return for `n == 0` would also work

Reply:
@elBoberido Well, because the checks are only needed under certain conditions :upside_down_face: 

1. When `n` is zero, no arithmetic operation is performed on the iterator, hence the check is not needed
2. When decrementing the iterator (negative `n`) the check is not needed because the iterator can go before the beginning of the `span`
3. When incrementing the iterator, the check is not needed because the iterator can go behind the end of `span`

Reply:
But this check is done for all conditions except `1` and then there is a potential assignment to a `nullptr` although it is `0`. In theory it should not cause any issue but I'm quite cautious with such things.

Reply:
Added.